espnet/nets/e2e_asr_transformer.py

```diff
@@ -0,0 +1,900 @@
+    scores is performed all heads together.
+
+    """
+
+    def __init__(self, n_units, h=8, dropout=0.1, self_attention=True, idim=0):
+        super(MultiHeadAttention, self).__init__()
+        idim = n_units if idim == 0 else idim
+
+        with self.init_scope():
+            if self_attention:
+                self.W_QKV = ConvolutionSentence(
+                    idim, n_units * 3, nobias=True,
+                    initialW=linear_init)
+            else:
+                self.W_Q = ConvolutionSentence(
+                    idim, n_units, nobias=True,
+                    initialW=linear_init)
+                self.W_KV = ConvolutionSentence(
+                    idim, n_units * 2, nobias=True,
+                    initialW=linear_init)
+            self.finishing_linear_layer = ConvolutionSentence(
+                n_units, n_units, nobias=True,
+                initialW=linear_init)
+        self.h = h
+        self.scale_score = 1. / (n_units // h) ** 0.5
+        self.dropout = dropout
+        self.is_self_attention = self_attention
+        self.n_units = n_units
+        self.idim = idim
+
+    def __call__(self, x, z=None, mask=None):
+        xp = self.xp
+        h = self.h
+
+        if self.is_self_attention:
+            Q, K, V = F.split_axis(self.W_QKV(x), 3, axis=1)
+        else:
+            Q = self.W_Q(x)
+            K, V = F.split_axis(self.W_KV(z), 2, axis=1)
+        batch, n_units, n_querys = Q.shape
+        _, _, n_keys = K.shape
+
+        # Calculate Attention Scores with Mask for Zero-padded Areas
+        # Perform Multi-head Attention using pseudo batching
+        # all together at once for efficiency
+
+        batch_Q = F.concat(F.split_axis(Q, h, axis=1), axis=0)
+        batch_K = F.concat(F.split_axis(K, h, axis=1), axis=0)
+        batch_V = F.concat(F.split_axis(V, h, axis=1), axis=0)
+        assert(batch_Q.shape == (batch * h, n_units // h, n_querys))
+        assert(batch_K.shape == (batch * h, n_units // h, n_keys))
+        assert(batch_V.shape == (batch * h, n_units // h, n_keys))
+        mask = xp.concatenate([mask] * h, axis=0)
+        batch_A = F.batch_matmul(batch_Q, batch_K, transa=True) \
+            * self.scale_score
+        batch_A = F.where(mask, batch_A, xp.full(batch_A.shape, -np.inf, 'f'))
+        batch_A = F.softmax(batch_A, axis=2)
+        batch_A = F.where(
+            xp.isnan(batch_A.data), xp.zeros(batch_A.shape, 'f'), batch_A)
+        assert(batch_A.shape == (batch * h, n_querys, n_keys))
+
+        # Calculate Weighted Sum
+        batch_A, batch_V = F.broadcast(
+            batch_A[:, None], batch_V[:, :, None])
+        batch_C = F.sum(batch_A * batch_V, axis=3)
+        assert(batch_C.shape == (batch * h, n_units // h, n_querys))
+        C = F.concat(F.split_axis(batch_C, h, axis=0), axis=1)
+        assert(C.shape == (batch, n_units, n_querys))
+        C = self.finishing_linear_layer(C)
+        return C
```

Comment:
for monitoring the soft alignment, it is nice to `return C, batch_A` and use `batch_A` instead of `C` in `calculate_all_attentions`